src/content/app-architecture/case-study/testing.md

```diff
@@ -0,0 +1,241 @@
+---
+title: Testing each layer
+short-title: Testing
+description: >
+  How to test an app that follows the recommended architecture.
+prev: 
+  title: Dependency injection
+  path: /app-architecture/case-study/dependency-injection
+---
+
+## Testing the UI layer
+
+One way to determine whether your architecture is sound is 
+considering how easy (or difficult) the application is to test. 
+Because ViewModels and views have well-defined inputs, 
+their dependencies can easily be mocked or faked, and unit tests are easily written.
+
+### ViewModel unit tests
+
+To test UI logic of the ViewModel, you should write unit tests that don't rely
+on Flutter libraries or testing frameworks.
+
+Repositories are a ViewModel's only dependencies 
+(unless you're implementing [use-cases][]), 
+and writing `mocks` or `fakes` of the repository is 
+the only setup you need to do. 
+In this example test, a fake called `FakeBookingRepository` is used.
+
+```dart
+// app/test/ui/home/widgets/home_screen_test.dart
+
+main() {
+  group('HomeViewModel tests', () {
+    test('Load bookings', () {
+      // HomeViewModel._load is called in the constructor of HomeViewModel
+      final viewModel = HomeViewModel(
+        bookingRepository: FakeBookingRepository()
+          ..createBooking(kBooking),
+        userRepository: FakeUserRepository(),
+      );
+
+      expect(viewModel.bookings.isNotEmpty, true);
+    });
+  });
+}
+```
+
+The [`FakeBookingRepository`][] class implements [`BookingRepository`][]. 
+In the [data layer section][] of this case-study, 
+the `BookingRepository` class is explained thoroughly.
+
+```dart
+// app/testing/fakes/repositories/fake_booking_repository.dart
+
+class FakeBookingRepository implements BookingRepository {
+  List<Booking> bookings = List.empty(growable: true);
+
+  @override
+  Future<Result<void>> createBooking(Booking booking) async {
+    bookings.add(booking);
+    return Result.ok(null);
+  }
+ // ...
+}
+```
+
+:::note
+If you're using this architecture with [use-cases][], these would
+similarly need to be faked.
+:::
+
+### View widget tests
+
+Once you've written tests for your ViewModel, 
+you've already created the fakes you need to write widget tests as well. 
+The following example shows how the `HomeScreen` widget tests 
+are set up using the `HomeViewModel` and needed repositories:
+
+```dart
+// app/test/ui/home/widgets/home_screen_test.dart
+
+main() {
```

Comment:
I recommend including the `void` return type for `main` methods, even in tests. As many developers have analysis configurations that end up requiring an explicit type.